From 3f5a6f00f2726dd793c156f5187f3460ed6fdec9 Mon Sep 17 00:00:00 2001
From: Fil Dunsky <filipp.dunsky@gmail.com>
Date: Tue, 3 Feb 2026 00:53:29 +0300
Subject: [PATCH] [draft] ramips: add support Securifi Almond 3S

WIP:
- LED
- Screen
- Zigbee
- Sound

1. Boot into initramfs:
In U-Boot press 4, set computer IP to 10.10.10.3, serve via tftp:
```
tftpboot 0x82000000 openwrt-24.10.5-ramips-mt7621-securifi_almond-3s-initramfs-kernel.bin
bootm 0x82000000
```
2. Backup all partitions via luci menu
3. Flash uboot-mt7621 by https://github.com/DragonBluep/uboot-mt7621
```
insmod mtd-rw i_want_a_brick=1
mtd write u-boot-mt7621.bin u-boot
```
4. Serve sysupgrade via luci, wait for 5 minutes.
5. Wait for 11 minutes more till the jffs2
initialize and enjoy OpenWrt!

Signed-off-by: Fil Dunsky <filipp.dunsky@gmail.com>
---
 .gitignore                                    |   1 -
 files/etc/atcommands.user                     |  15 ++
 files/etc/banner                              |   9 +
 files/etc/config/3ginfo                       |   4 +
 files/etc/mdm.sh                              |   6 +
 files/etc/opkg/distfeeds.conf                 |   7 +
 files/etc/rc.local                            |   5 +
 files/etc/uci-defaults/99-Almond.sh           |  11 +
 package/base-files/Makefile                   |   2 +-
 package/luci-app-3ginfo-lite                  |   1 +
 package/luci-app-modemband                    |   1 +
 package/luci-app-sms-tool-js                  |   1 +
 package/luci-theme-proton2025                 |   1 +
 .../ramips/dts/mt7621_securifi_almond-3s.dts  | 205 ++++++++++++++++++
 target/linux/ramips/image/mt7621.mk           |  13 ++
 .../mt7621/base-files/etc/board.d/02_network  |   5 +
 16 files changed, 285 insertions(+), 2 deletions(-)
 create mode 100644 files/etc/atcommands.user
 create mode 100644 files/etc/banner
 create mode 100644 files/etc/config/3ginfo
 create mode 100755 files/etc/mdm.sh
 create mode 100644 files/etc/opkg/distfeeds.conf
 create mode 100755 files/etc/rc.local
 create mode 100644 files/etc/uci-defaults/99-Almond.sh
 create mode 160000 package/luci-app-3ginfo-lite
 create mode 160000 package/luci-app-modemband
 create mode 160000 package/luci-app-sms-tool-js
 create mode 160000 package/luci-theme-proton2025
 create mode 100644 target/linux/ramips/dts/mt7621_securifi_almond-3s.dts

diff --git a/.gitignore b/.gitignore
index ad0475591915f4..964c918259f336 100644
--- a/.gitignore
+++ b/.gitignore
@@ -13,7 +13,6 @@
 /logs
 /feeds
 /feeds.conf
-/files
 /target/linux/feeds
 /overlay
 /package/feeds
diff --git a/files/etc/atcommands.user b/files/etc/atcommands.user
new file mode 100644
index 00000000000000..dc253a138fc9e6
--- /dev/null
+++ b/files/etc/atcommands.user
@@ -0,0 +1,15 @@
+Check signal strength > AT+CSQ;AT+CSQ
+Get the temperature of MT > AT+QTEMP;AT+QTEMP
+Current band in use > AT+QNWINFO;AT+QNWINFO
+Band 1+3+7+8+20+38+40+41 > AT+QCFG="band",0,1A0000800C5;AT+QCFG="band",0,1A0000800C5
+Band 1+3+7+38 > AT+QCFG="band",0,2000000045;AT+QCFG="band",0,2000000045
+Band 3+7+38 > AT+QCFG="band",0,2000000044;AT+QCFG="band",0,2000000044
+Band 3+7 > AT+QCFG="band",0,0000000044;AT+QCFG="band",0,0000000044
+Band 7 > AT+QCFG="band",0,0000000040;AT+QCFG="band",0,0000000040
+Band 38 > AT+QCFG="band",0,2000000000;AT+QCFG="band",0,2000000000
+Reboot modem  >  AT+CFUN=1,1;AT+CFUN=1,1
+QMI/PPP/Default > AT+QCFG="usbnet",0;AT+QCFG="usbnet",0
+Network search LTE/WCDMA/GSM > AT+QCFG=“nwscanseq”,040301;AT+QCFG=“nwscanseq”,040301
+4G-LTE only > AT+QCFG="nwscanmode",3,1;AT+QCFG="nwscanmode",3,1
+Scan all modes > AT+QCFG="nwscanmode",0,1;AT+QCFG="nwscanmode",0,1
+Change IMEI > AT+EGMR=1,7,"000000000000000";AT+EGMR=1,7,"000000000000000"
diff --git a/files/etc/banner b/files/etc/banner
new file mode 100644
index 00000000000000..252d2161aa9278
--- /dev/null
+++ b/files/etc/banner
@@ -0,0 +1,9 @@
+  _______ __                          __ _____ 
+ |   _   |  |.--------.-----.-----.--|  |__   |
+ |       |  ||        |  _  |     |  _  |__   |
+ |___|___|  ||__|__|__|_____|__|__|_____|_____|
+         |____| S E C O N D   L I F E                                
+ ----------------------------------------------
+      Securifi Almond 3S, OpenWrt 24.10.5
+ ----------------------------------------------
+
diff --git a/files/etc/config/3ginfo b/files/etc/config/3ginfo
new file mode 100644
index 00000000000000..2ed091821ad971
--- /dev/null
+++ b/files/etc/config/3ginfo
@@ -0,0 +1,4 @@
+config 3ginfo
+        option website ''
+        option network 'modem'
+        option device '/dev/ttyUSB2'
diff --git a/files/etc/mdm.sh b/files/etc/mdm.sh
new file mode 100755
index 00000000000000..a18958a03416ca
--- /dev/null
+++ b/files/etc/mdm.sh
@@ -0,0 +1,6 @@
+#!/bin/sh
+#Toggle modem power off
+
+echo 0 > /sys/class/gpio/modem_reset/value
+sleep 5
+echo 1 > /sys/class/gpio/modem_reset/value
diff --git a/files/etc/opkg/distfeeds.conf b/files/etc/opkg/distfeeds.conf
new file mode 100644
index 00000000000000..4d9592c6b118ba
--- /dev/null
+++ b/files/etc/opkg/distfeeds.conf
@@ -0,0 +1,7 @@
+src/gz openwrt_core https://downloads.openwrt.org/releases/24.10.5/targets/ramips/mt7621/packages
+src/gz openwrt_base https://downloads.openwrt.org/releases/24.10.5/packages/mipsel_24kc/base
+src/gz openwrt_kmods https://downloads.openwrt.org/releases/24.10.5/targets/ramips/mt7621/kmods/6.6.119-1-6c0cbfffdf5543d41b1de30e3a9c928d
+src/gz openwrt_luci https://downloads.openwrt.org/releases/24.10.5/packages/mipsel_24kc/luci
+src/gz openwrt_packages https://downloads.openwrt.org/releases/24.10.5/packages/mipsel_24kc/packages
+src/gz openwrt_routing https://downloads.openwrt.org/releases/24.10.5/packages/mipsel_24kc/routing
+src/gz openwrt_telephony https://downloads.openwrt.org/releases/24.10.5/packages/mipsel_24kc/telephony
diff --git a/files/etc/rc.local b/files/etc/rc.local
new file mode 100755
index 00000000000000..3bb132634d13e3
--- /dev/null
+++ b/files/etc/rc.local
@@ -0,0 +1,5 @@
+# Move USB to Core1t2, WiFi 2.4GHz and 5GHz to Core2t1 and Core2t1 
+echo 2 > "/proc/irq/29/smp_affinity"
+echo 4 > "/proc/irq/31/smp_affinity"
+echo 8 > "/proc/irq/32/smp_affinity"
+exit 0
diff --git a/files/etc/uci-defaults/99-Almond.sh b/files/etc/uci-defaults/99-Almond.sh
new file mode 100644
index 00000000000000..dd8d393e9e286c
--- /dev/null
+++ b/files/etc/uci-defaults/99-Almond.sh
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+uci set system.@system[0].hostname='Almond'
+uci set system.@system[0].timezone='MSK-3'
+uci set system.@system[0].zonename='Europe/Moscow'
+uci commit system
+echo 'Almond' > /proc/sys/kernel/hostname
+/etc/init.d/system reload
+
+exit 0
+
diff --git a/package/base-files/Makefile b/package/base-files/Makefile
index c78c073699f2dc..b601bd6121c29a 100644
--- a/package/base-files/Makefile
+++ b/package/base-files/Makefile
@@ -13,7 +13,7 @@ include $(INCLUDE_DIR)/feeds.mk
 
 PKG_NAME:=base-files
 PKG_FLAGS:=nonshared
-PKG_RELEASE:=$(COMMITCOUNT)
+PKG_RELEASE:=1668
 
 PKG_FILE_DEPENDS:=$(PLATFORM_DIR)/ $(GENERIC_PLATFORM_DIR)/base-files/
 PKG_BUILD_DEPENDS:=usign/host ucert/host
diff --git a/package/luci-app-3ginfo-lite b/package/luci-app-3ginfo-lite
new file mode 160000
index 00000000000000..6861ee9e217591
--- /dev/null
+++ b/package/luci-app-3ginfo-lite
@@ -0,0 +1 @@
+Subproject commit 6861ee9e2175919848494edbf1c534bbc1e45937
diff --git a/package/luci-app-modemband b/package/luci-app-modemband
new file mode 160000
index 00000000000000..85fe3d6b9d6e7c
--- /dev/null
+++ b/package/luci-app-modemband
@@ -0,0 +1 @@
+Subproject commit 85fe3d6b9d6e7cb2c332fd85589b7f9cb0902507
diff --git a/package/luci-app-sms-tool-js b/package/luci-app-sms-tool-js
new file mode 160000
index 00000000000000..16b0e7b08e7096
--- /dev/null
+++ b/package/luci-app-sms-tool-js
@@ -0,0 +1 @@
+Subproject commit 16b0e7b08e70965ac271b476134bf7459ab9b7a1
diff --git a/package/luci-theme-proton2025 b/package/luci-theme-proton2025
new file mode 160000
index 00000000000000..bf273090c04ea0
--- /dev/null
+++ b/package/luci-theme-proton2025
@@ -0,0 +1 @@
+Subproject commit bf273090c04ea0d126c5edac2320f1aaaf95098a
diff --git a/target/linux/ramips/dts/mt7621_securifi_almond-3s.dts b/target/linux/ramips/dts/mt7621_securifi_almond-3s.dts
new file mode 100644
index 00000000000000..0456ceafc6525c
--- /dev/null
+++ b/target/linux/ramips/dts/mt7621_securifi_almond-3s.dts
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/*
+ * Copyright (c) 2026
+ * Author: Fil Dunsky <filipp.dunsky@gmail.com>
+ */
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	model = "Securifi Almond 3S";
+	compatible = "securifi,almond-3s", "mediatek,mt7621-soc";
+
+	chosen {
+		bootargs = "console=ttyS0,57600n8";
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 32 GPIO_ACTIVE_LOW>;
+			debounce-interval = <20>;
+		};
+		
+		tamper {
+			label = "tamper";
+			linux,code = <BTN_0>;
+			gpios = <&gpio 28 GPIO_ACTIVE_LOW>;
+			debounce-interval = <20>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		display_power {
+			label = "display:power";
+			gpios = <&gpio 31 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		modem-reset {
+			gpio-export,name = "modem_reset";
+			gpio-export,output = <1>;
+			gpios = <&gpio 33 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&i2c {
+	status = "okay";
+};
+
+&gdma {
+	status = "okay";
+};
+
+&i2s {
+	status = "okay";
+	#sound-dai-cells = <0>;
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 { /* MT25QL512AB */
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <52000000>;
+		broken-flash-reset;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+			};
+
+			partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+
+				nvmem-layout {
+					compatible = "fixed-layout";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					eeprom_factory_0: eeprom@0 {
+						reg = <0x0 0x200>;
+					};
+
+					eeprom_factory_8000: eeprom@8000 {
+						reg = <0x8000 0x200>;
+					};
+					
+					mac_lan: macaddr@28 {
+						reg = <0x28 0x6>;
+					};
+
+					mac_wan: macaddr@2e {
+						reg = <0x2e 0x6>;
+					};					
+				};
+			};
+			
+			partition@50000 {
+					compatible = "openwrt,uimage", "denx,uimage";
+					label = "firmware";
+					reg = <0x0050000 0x3fb0000>;
+			};
+		};
+	};
+};
+
+&pcie {
+	status = "okay";
+};
+
+&pcie0 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		ieee80211-freq-limit = <5000000 6000000>;
+
+		nvmem-cells = <&eeprom_factory_8000>;
+		nvmem-cell-names = "eeprom";
+	};
+};
+
+&pcie1 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		ieee80211-freq-limit = <2400000 2500000>;
+
+		nvmem-cells = <&eeprom_factory_0>;
+		nvmem-cell-names = "eeprom";
+	};
+};
+
+&gmac0 {
+	nvmem-cells = <&mac_lan>;
+	nvmem-cell-names = "mac-address";
+};
+
+&ethernet {
+	pinctrl-0 = <&mdio_pins>, <&rgmii1_pins>;
+};
+
+&state_default {
+	gpio {
+		groups = "rgmii2";
+		function = "gpio";
+	};
+};
+
+&switch0 {
+	ports {
+		port@0 {
+			status = "okay";
+			label = "wan";
+			nvmem-cells = <&mac_wan>;
+			nvmem-cell-names = "mac-address";
+		};
+        
+		port@1 {
+			status = "okay";
+			label = "lan2";
+		};
+
+		port@2 {
+			status = "okay";
+			label = "lan1";
+		};
+	};
+};
+
+&uartlite2 {
+	status = "okay";
+	current-speed = <57600>;
+};
+
+&uartlite3 {
+	status = "okay";
+	current-speed = <57600>;
+};
diff --git a/target/linux/ramips/image/mt7621.mk b/target/linux/ramips/image/mt7621.mk
index 4184977b27e11a..6f7aa79f2186b0 100644
--- a/target/linux/ramips/image/mt7621.mk
+++ b/target/linux/ramips/image/mt7621.mk
@@ -2564,6 +2564,19 @@ define Device/samknows_whitebox-v8
 endef
 TARGET_DEVICES += samknows_whitebox-v8
 
+define Device/securifi_almond-3s
+  $(Device/dsa-migration)
+  DEVICE_VENDOR := Securifi
+  DEVICE_MODEL := Almond
+  DEVICE_VARIANT := 3S
+  IMAGE_SIZE := 65216k
+  IMAGE/sysupgrade.bin := append-kernel | pad-to 64k | append-rootfs | \
+        pad-rootfs | check-size | append-metadata
+  DEVICE_PACKAGES := kmod-mt76x2 kmod-usb3 -uboot-envtools \
+  	kmod-usb-net-qmi-wwan kmod-usb-serial-option uqmi
+endef
+TARGET_DEVICES += securifi_almond-3s
+
 define Device/sercomm_na502
   $(Device/nand)
   $(Device/uimage-lzma-loader)
diff --git a/target/linux/ramips/mt7621/base-files/etc/board.d/02_network b/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
index 557755064a82e3..607370c62fa084 100644
--- a/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
@@ -108,6 +108,11 @@ ramips_setup_interfaces()
 	xiaomi,mi-router-4a-gigabit-v2)
 		ucidef_set_interfaces_lan_wan "lan1 lan2" "wan"
 		;;
+	securifi,almond-3s)
+		ucidef_set_interfaces_lan_wan "lan1 lan2" "wan"
+		ucidef_set_interface "modem" device "/dev/cdc-wdm0" protocol "qmi"
+		uci add_list firewall.@zone[1].network='modem'
+		;;
 	bolt,arion)
 		ucidef_set_interfaces_lan_wan "lan" "wan"
 		ucidef_set_interface "eth_data" device "modem.103" protocol "static" ipaddr "10.22.127.222" netmask "255.255.255.255"